src/collections.rs

```diff
@@ -155,3 +161,149 @@ impl From<&[u8]> for z_bytes_t {
+        z_buffer_null()
+    }
+}
+impl From<ZBuf> for z_owned_buffer_t {
+    fn from(value: ZBuf) -> Self {
+        let value = match value.contiguous() {
+            std::borrow::Cow::Borrowed(_) => value,
+            std::borrow::Cow::Owned(value) => value.into(),
+        };
+        unsafe { core::mem::transmute(Some(value)) }
+    }
+}
+impl From<Option<ZBuf>> for z_owned_buffer_t {
+    fn from(value: Option<ZBuf>) -> Self {
+        match value {
+            Some(value) => value.into(),
+            None => z_buffer_null(),
+        }
+    }
+}
+impl core::ops::Deref for z_owned_buffer_t {
+    type Target = Option<ZBuf>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+impl core::ops::DerefMut for z_owned_buffer_t {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+
+/// The gravestone value for `z_owned_buffer_t`.
+#[no_mangle]
+pub extern "C" fn z_buffer_null() -> z_owned_buffer_t {
+    unsafe { core::mem::transmute(None::<ZBuf>) }
+}
+
+/// Decrements the buffer's reference counter, destroying it if applicable.
+///
+/// `buffer` will be reset to `z_buffer_null`, preventing UB on double-frees.
+#[no_mangle]
+pub extern "C" fn z_buffer_drop(buffer: &mut z_owned_buffer_t) {
+    core::mem::drop(buffer.take())
+}
+
+/// Returns `true` if the buffer is in a valid state.
+#[no_mangle]
+pub extern "C" fn z_buffer_check(buffer: &z_owned_buffer_t) -> bool {
+    buffer.is_some()
+}
+
+/// Loans the buffer, allowing you to call functions that only need a loan of it.
+#[no_mangle]
+pub extern "C" fn z_buffer_loan(buffer: &z_owned_buffer_t) -> z_buffer_t {
+    buffer.as_ref().into()
+}
+
+/// A loan of a `z_owned_buffer_t`.
+///
+/// As it is a split buffer, it may contain more than one slice. It's number of slices is returned by `z_buffer_slice_count`.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct z_buffer_t<'a> {
+    _inner: &'a (),
+}
+impl_guarded_transmute!(Option<&'a ZBuf>, z_buffer_t<'a>, 'a);
+impl<'a> From<z_buffer_t<'a>> for Option<&'a ZBuf> {
+    fn from(value: z_buffer_t) -> Self {
+        unsafe { core::mem::transmute(value) }
+    }
+}
+
+/// Increments the buffer's reference count, returning an owned version of the buffer.
+#[no_mangle]
+pub extern "C" fn z_buffer_clone(buffer: z_buffer_t) -> z_owned_buffer_t {
+    unsafe { Some(core::mem::transmute::<_, &ZBuf>(buffer).clone()).into() }
+}
+
+/// Returns the payload of the buffer if it is contiguous, aliasling it.
+///
+/// If the payload was not contiguous in memory, `z_bytes_null` will be returned instead.
+#[no_mangle]
+pub extern "C" fn z_buffer_payload(buffer: z_buffer_t) -> z_bytes_t {
+    let Some(buffer): Option<&ZBuf> = buffer.into() else {
+        return z_bytes_null();
+    };
+    match buffer.contiguous() {
+        std::borrow::Cow::Borrowed(buffer) => buffer.into(),
+        std::borrow::Cow::Owned(_) => z_bytes_null(),
```

Comment:
if we get the Owned enum member, it means we already allocated and copied element into a vector. It seems to be a pity not to return the data in this case.

Reply:
The core problem here is that `z_bytes_t` is a borrowed type according to our convention: it may never be owned by the user.